File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/NoteDao.java
package de.westnordost.streetcomplete.data.osmnotes;


import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;

import javax.inject.Inject;

import de.westnordost.streetcomplete.util.Serializer;
import de.westnordost.osmapi.map.data.OsmLatLon;
import de.westnordost.osmapi.notes.Note;

public class NoteDao
{
	private final SQLiteOpenHelper dbHelper;
	private final Serializer serializer;

	private final SQLiteStatement insert;

	@Inject public NoteDao(SQLiteOpenHelper dbHelper, Serializer serializer)
	{
		this.dbHelper = dbHelper;
		this.serializer = serializer;

		String sql = "INSERT OR REPLACE INTO " + NoteTable.NAME + " ("+
				NoteTable.Columns.ID+","+
				NoteTable.Columns.LATITUDE+","+
				NoteTable.Columns.LONGITUDE+","+
				NoteTable.Columns.STATUS+","+
				NoteTable.Columns.CREATED+","+
				NoteTable.Columns.CLOSED+","+
				NoteTable.Columns.COMMENTS+
				") values (?,?,?,?,?,?,?);";
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		insert = db.compileStatement(sql);
	}

	public void putAll(Collection<Note> notes)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		db.beginTransaction();

		for(Note note : notes)
		{
			executeInsert(note);
		}

		db.setTransactionSuccessful();
		db.endTransaction();
	}

	public void put(Note note)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		db.beginTransaction();
		executeInsert(note);
		db.setTransactionSuccessful();
		db.endTransaction();
	}

	private void executeInsert(Note note)
	{
		insert.bindLong(1, note.id);
		insert.bindDouble(2, note.position.getLatitude());
		insert.bindDouble(3, note.position.getLongitude());
		insert.bindString(4, note.status.name());
		insert.bindLong(5, note.dateCreated.getTime());
		if(note.dateClosed != null)
		{
			insert.bindLong(6, note.dateClosed.getTime());
		}
		else
		{
			insert.bindNull(6);
		}
		insert.bindBlob(7, serializer.toBytes(note.comments));

		insert.executeInsert();
		insert.clearBindings();
	}

	public Note get(long id)
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();
		String where = NoteTable.Columns.ID + " = " + id;

		try (Cursor cursor = db.query(NoteTable.NAME, null, where,
				null, null, null, null, "1"))
		{
			if (!cursor.moveToFirst()) return null;
			return createObjectFrom(serializer, cursor);
		}
	}

	public boolean delete(long id)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		return db.delete(NoteTable.NAME, NoteTable.Columns.ID + " = " + id, null) == 1;
	}

	static Note createObjectFrom(Serializer serializer, Cursor cursor)
	{
		int colNoteId = cursor.getColumnIndexOrThrow(NoteTable.Columns.ID),
			colLat = cursor.getColumnIndexOrThrow(NoteTable.Columns.LATITUDE),
			colLon = cursor.getColumnIndexOrThrow(NoteTable.Columns.LONGITUDE),
			colStatus = cursor.getColumnIndexOrThrow(NoteTable.Columns.STATUS),
			colCreated = cursor.getColumnIndexOrThrow(NoteTable.Columns.CREATED),
			colClosed = cursor.getColumnIndexOrThrow(NoteTable.Columns.CLOSED),
			colComments = cursor.getColumnIndexOrThrow(NoteTable.Columns.COMMENTS);

		Note note = new Note();
		note.id = cursor.getLong(colNoteId);
		note.position = new OsmLatLon(cursor.getDouble(colLat), cursor.getDouble(colLon));
		note.dateCreated = new Date(cursor.getLong(colCreated));
		if(!cursor.isNull(colClosed))
		{
			note.dateClosed = new Date(cursor.getLong(colClosed));
		}
		note.status = Note.Status.valueOf(cursor.getString(colStatus));
		note.comments = serializer.toObject(cursor.getBlob(colComments), ArrayList.class);

		return note;
	}

	public int deleteUnreferenced()
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		String where = NoteTable.Columns.ID + " NOT IN ( " +
				"SELECT " + OsmNoteQuestTable.Columns.NOTE_ID + " FROM " + OsmNoteQuestTable.NAME +
				")";

		return db.delete(NoteTable.NAME, where, null);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/NoteDao.kt
package de.westnordost.streetcomplete.data.osmnotes

import de.westnordost.streetcomplete.data.CursorPosition
import de.westnordost.streetcomplete.data.Database
import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.CLOSED
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.COMMENTS
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.CREATED
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.ID
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.LAST_SYNC
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.LATITUDE
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.LONGITUDE
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.Columns.STATUS
import de.westnordost.streetcomplete.data.osmnotes.NoteTable.NAME
import de.westnordost.streetcomplete.util.ktx.nowAsEpochMilliseconds
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Stores OSM notes */
class NoteDao(private val db: Database) {
    fun put(note: Note) {
        db.replace(NAME, note.toPairs())
    }

    fun get(id: Long): Note? =
        db.queryOne(NAME, where = "$ID = $id") { it.toNote() }

    fun delete(id: Long): Boolean =
        db.delete(NAME, "$ID = $id") == 1

    fun putAll(notes: Collection<Note>) {
        if (notes.isEmpty()) return

        db.replaceMany(NAME,
            arrayOf(ID, LATITUDE, LONGITUDE, STATUS, CREATED, CLOSED, COMMENTS, LAST_SYNC),
            notes.map { arrayOf(
                it.id,
                it.position.latitude,
                it.position.longitude,
                it.status.name,
                it.timestampCreated,
                it.timestampClosed,
                Json.encodeToString(it.comments),
                nowAsEpochMilliseconds()
            ) }
        )
    }

    fun getAll(bbox: BoundingBox): List<Note> =
        db.query(NAME, where = inBoundsSql(bbox)) { it.toNote() }

    fun getAllPositions(bbox: BoundingBox): List<LatLon> =
        db.query(NAME,
            columns = arrayOf(LATITUDE, LONGITUDE),
            where = inBoundsSql(bbox),
        ) { LatLon(it.getDouble(LATITUDE), it.getDouble(LONGITUDE)) }

    fun getAll(ids: Collection<Long>): List<Note> {
        if (ids.isEmpty()) return emptyList()
        return db.query(NAME, where = "$ID IN (${ids.joinToString(",")})") { it.toNote() }
    }

    fun getIdsOlderThan(timestamp: Long, limit: Int? = null): List<Long> {
        if (limit != null && limit <= 0) return emptyList()
        else return db.query(NAME,
            columns = arrayOf(ID),
            where = "$LAST_SYNC < $timestamp",
            limit = limit?.toString()
        ) { it.getLong(ID) }
    }

    fun deleteAll(ids: Collection<Long>): Int {
        if (ids.isEmpty()) return 0
        return db.delete(NAME, "$ID IN (${ids.joinToString(",")})")
    }

    fun clear() {
        db.delete(NAME)
    }

    private fun Note.toPairs() = listOf(
        ID to id,
        LATITUDE to position.latitude,
        LONGITUDE to position.longitude,
        STATUS to status.name,
        CREATED to timestampCreated,
        CLOSED to timestampClosed,
        COMMENTS to Json.encodeToString(comments),
        LAST_SYNC to nowAsEpochMilliseconds()
    )

    private fun CursorPosition.toNote() = Note(
        LatLon(getDouble(LATITUDE), getDouble(LONGITUDE)),
        getLong(ID),
        getLong(CREATED),
        getLongOrNull(CLOSED),
        Note.Status.valueOf(getString(STATUS)),
        Json.decodeFromString(getString(COMMENTS))
    )

    private fun inBoundsSql(bbox: BoundingBox): String = """
        ($LATITUDE BETWEEN ${bbox.min.latitude} AND ${bbox.max.latitude}) AND
        ($LONGITUDE BETWEEN ${bbox.min.longitude} AND ${bbox.max.longitude})
    """.trimIndent()
}
